File: supabase/functions/_shared/deepseek.ts
const DEEPSEEK_CHAT_COMPLETIONS_URL = 'https://api.deepseek.com/chat/completions';

interface DeepSeekJsonRequest {
  systemPrompt: string;
  userPrompt: string;
  maxTokens: number;
  maxAttempts?: number;
  requestLabel: string;
}

interface DeepSeekChatResponse {
  choices?: Array<{
    message?: {
      content?: string | null;
    };
  }>;
  error?: {
    message?: string;
  };
}

export async function callDeepSeekJson({
  systemPrompt,
  userPrompt,
  maxTokens,
  maxAttempts = 3,
  requestLabel,
}: DeepSeekJsonRequest): Promise<unknown> {
  const apiKey = Deno.env.get('DEEPSEEK_API_KEY');
  if (!apiKey) {
    throw new Error('Missing DEEPSEEK_API_KEY secret');
  }

  let lastError: Error | null = null;

  for (let attempt = 1; attempt <= maxAttempts; attempt += 1) {
    try {
      const response = await fetch(DEEPSEEK_CHAT_COMPLETIONS_URL, {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${apiKey}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          model: 'deepseek-chat',
          messages: [
            { role: 'system', content: systemPrompt },
            { role: 'user', content: userPrompt },
          ],
          max_tokens: maxTokens,
          temperature: 0.2,
          response_format: { type: 'json_object' },
        }),
      });

      if (!response.ok) {
        const bodyText = await response.text();
        throw new Error(`DeepSeek API error (${response.status}): ${bodyText}`);
      }

      const data = await response.json() as DeepSeekChatResponse;
      const rawContent = data.choices?.[0]?.message?.content;
      if (!rawContent || !rawContent.trim()) {
        throw new Error('DeepSeek returned empty content');
      }

      const cleaned = rawContent
        .replace(/^```(?:json)?\s*/i, '')
        .replace(/\s*```$/i, '')
        .trim();

      return JSON.parse(cleaned);
    } catch (error) {
      lastError = error instanceof Error ? error : new Error(String(error));
      console.warn(`[${requestLabel}] attempt ${attempt} failed:`, lastError.message);

      if (attempt < maxAttempts) {
        await sleep(500 * 2 ** (attempt - 1));
      }
    }
  }

  throw lastError ?? new Error('DeepSeek request failed after retries');
}

function sleep(ms: number) {
  return new Promise<void>((resolve) => setTimeout(resolve, ms));
}

File: backend/app/services/deepseek.py
"""Port of supabase/functions/_shared/deepseek.ts"""

import re
import time
import json
import logging

import httpx
from fastapi import HTTPException

from ..config import settings

DEEPSEEK_URL = "https://api.deepseek.com/chat/completions"

logger = logging.getLogger(__name__)


def call_deepseek_json(
    *,
    system_prompt: str,
    user_prompt: str,
    max_tokens: int,
    max_attempts: int = 3,
    request_label: str = "deepseek",
) -> object:
    """Call the DeepSeek chat API and return a parsed JSON object.

    Retries with exponential backoff on failure. Strips markdown fences from
    the response before parsing.
    """
    last_error: Exception | None = None

    for attempt in range(1, max_attempts + 1):
        try:
            response = httpx.post(
                DEEPSEEK_URL,
                headers={
                    "Authorization": f"Bearer {settings.DEEPSEEK_API_KEY}",
                    "Content-Type": "application/json",
                },
                json={
                    "model": "deepseek-chat",
                    "messages": [
                        {"role": "system", "content": system_prompt},
                        {"role": "user", "content": user_prompt},
                    ],
                    "max_tokens": max_tokens,
                    "temperature": 0.2,
                    "response_format": {"type": "json_object"},
                },
                timeout=120.0,
            )

            if not response.is_success:
                raise RuntimeError(
                    f"DeepSeek API error ({response.status_code}): {response.text}"
                )

            data = response.json()
            raw_content: str | None = (
                data.get("choices", [{}])[0].get("message", {}).get("content")
            )
            if not raw_content or not raw_content.strip():
                raise RuntimeError("DeepSeek returned empty content")

            cleaned = re.sub(r"^```(?:json)?\s*", "", raw_content, flags=re.IGNORECASE)
            cleaned = re.sub(r"\s*```$", "", cleaned).strip()

            return json.loads(cleaned)

        except Exception as exc:  # noqa: BLE001
            last_error = exc
            logger.warning("[%s] attempt %d failed: %s", request_label, attempt, exc)
            if attempt < max_attempts:
                time.sleep(0.5 * (2 ** (attempt - 1)))

    raise HTTPException(
        status_code=503,
        detail=f"DeepSeek request failed after {max_attempts} attempts: {last_error}",
    )
